[util] Add ability to declare parameter values from top

tjaychen opened the pull request:
[util] Add ability to declare parameter level directly from top_earlgrey.hjson

Addresses #5227

- This will allow us to control what the ASIC variant should look like from
  inside top_earlgrey, while still allowing top level overrides for FPGA.

- This will be particularly useful for rv_core_ibex templating

Signed-off-by: Timothy Chen <timothytim@google.com>

msfschaffner commented on util/topgen/lib.py:
```diff
@@ -495,3 +495,15 @@ def make_bit_concatenation(sig_name: str,
+
+def param_value (ip_param, module_params_decl):
+    '''Return param value of an exposed parameter.
+       If there are no explicitly declared value under module, use default.
```
nit: value -> value**s**

rswarbrick replied:
Actually, I think it should probably be "If there is no explicitly..."? Also, could you add a newline before the second sentence? The usual style for Python docstrings is like Git commit messages: a single line title, then zero or more paragraphs separated by blank lines.

tjaychen replied:
ah yes, the next line seems to be singular and this line plural for some reason..let me switch to singular for both.

msfschaffner commented on util/topgen/merge.py:
```diff
@@ -119,6 +124,13 @@ def elaborate_instance(instance, block: IpBlock):
+    for decl in instance["param_decl"].keys():
+        if not any(decl in param['name'] for param in instance["param_list"]):
+            log.error("{} is not a valid parameter of {} that can be "
+                      "set from top level".format(decl, block.name))
```
how about we directly override the `default` values in this function? by doing that, we could do away with the additional helper function in the template (I've done something similar with the `MemSize` parameter in #7267). 

for example, we could add the following `elif` leg to the `if isinstance(param, RandParameter):` statement on line 109 above:

```python
        # if this exposed parameter is listed in the `param_decl` dict, 
        # override its default value.
        elif param.name in instance["param_decl"]:
            new_param['default'] = instance["param_decl"][param.name]
            # remove the parameter from the accounting dict
            del param_decl_accounting[param.name]
```

`param_decl_accounting` would have to be initialized with 
```python
param_decl_accounting = instance["param_decl"].copy()
```
before the loop on line 77. this accounting dict can then be used to check that all parameters have been matched after the loop on line 126
```python
    # we should not have left-over declarations here.
    for decl in param_decl:
         log.error("{} is not a valid parameter of {} that can be "
                   "set from top level".format(decl, block.name))
```

tjaychen replied:
yeah i thought about overriding also, but it felt a bit like contaminating the definition of "default" by doing that. 
I don't have strong feelings one way or the other though, so if others also feel like overriding is the better option, I'm happy to do that.

msfschaffner replied:
Yeah I know, `default` feels a bit like a misnomer in this case... I think it has originally been named `default` since this is the value that is assigned in the `top_earlgrey` module parameter definition, which can later be overridden in RTL wherever the `top_earlgrey` module is instantiated.

Given that we already override these values here for the `RndCnst` parameters, shall we just keep the override mechanism here to make this a bit more contained?

tjaychen replied:
sorry when you say keep it here, you mean override in merge and not at top level? 
I'm at moment inclined towards  top level out of...laziness :D

tjaychen replied:
after looking at this more closely, i think you are right. It is better to change it here since the `Rand` case already does the same.
I've done mostly what you suggested, with a slight change (didn't use copy..since i can't quite tell if this puts some of the underlying objects at risk if we added them in the future).

msfschaffner reviewed (commented):
This looks useful! I only have a suggestion on where we could override the default values.

rswarbrick commented on util/topgen/lib.py:
```diff
@@ -495,3 +495,15 @@ def make_bit_concatenation(sig_name: str,
     return ''.join(acc)
+
+
+def param_value (ip_param, module_params_decl):
```
Nit: PEP8 style (which we claim to follow) has no space before the opening paren

tjaychen replied:
thanks! i should have noticed the lint failure.

rswarbrick commented on util/topgen/lib.py:
```diff
@@ -495,3 +495,15 @@ def make_bit_concatenation(sig_name: str,
+       If there is an explicitly declared value, use that instead.
+
+    '''
+    if (ip_param["name"] in module_params_decl.keys()):
```
Nit: You don't need the outer parens here. You can also avoid the multiple return statements with something like this (which also avoids double lookups, not that we really care about the performance impact):
```python3
    mod_val = module_params_decl.get(ip_param["name"])
    return mod_val if mod_val is not None else ip_param["default"]
```

tjaychen replied:
thanks! i didnt even know there was a `get` method, that's super helpful!

rswarbrick reviewed (approved):
A couple of nits about Python style, but this looks good. No strong feelings about the question that Michael raised: I'm happy either way.

tjaychen commented on hw/top_earlgrey/data/top_earlgrey.hjson:
```diff
@@ -566,8 +566,8 @@
     },
-    { name: "kmac"
-      type: "kmac"
+    { name: "kmac",
+      type: "kmac",
```
ah sorry, this is mostly because of the editor i'm using. It doesn't indent the next line nicely unless the previous line has a comma. Let me remove this.

tjaychen replied:
just curious, what editor do people use? i'm a vanilla emacs user..but wondering if i should be updating...

rswarbrick replied:
Emacs here too. For hjson, I've given up on clever modes, so tend to put it in C mode for the paren matching and avoid hitting tab... For languages that anyone else uses, Emacs is brilliant.

tjaychen commented:
thanks all for the review, i will merge this for now.
Just as a summary, I changed to @msfschaffner suggestion of placing it all in `merge.py` and removed the helper function under `lib.py`.
